source/connections-survive-step-down/tests/README.rst

```diff
@@ -71,8 +71,19 @@ Perform the following operations:
 - If the driver implements the `CMAP`_ specification, verify that no new `PoolClearedEvent`_ has been
   published. Otherwise verify that `connections.totalCreated`_ in `serverStatus`_ has not changed.
 
+**Note:** The "replSetStepDown" command often fails with the following
+transient error (see `SERVER-48154`_)::
 
+  {
+    "ok" : 0,
+    "errmsg" : "Unable to acquire X lock on '{4611686018427387905: ReplicationStateTransition, 1}' within 1000ms. opId: 922, op: conn30, connId: 30.",
+    "code" : 24,
+    "codeName" : "LockTimeout",
+  }
 
+When running the "replSetStepDown" command, drivers MUST retry until the
+command succeeds. The number of retries should be limited to avoid an infinite
```

Comment:
What should the limit be?

Reply:
I think it needs to be at least 10 seconds total given the description in SERVER-48154:
> If there are hundreds of idents to drop, the dropper can take 10 seconds or more, during which time stepdown and shutdown are blocked.

I added "For example, the Python driver uses a 10 second retry period."

Reply:
Instead of doing a retry, can we increase the time value in the `replSetStepdown` command? Something like: `{replSetStepDown: 60, force: true}`?

The server test mentioned in https://jira.mongodb.org/browse/SERVER-48107 uses a time period of 1. The stepdown tests use a [time period of 5](https://github.com/mongodb/specifications/blob/master/source/connections-survive-step-down/tests/README.rst#getmore-iteration).

It appears that `replSetStepDown` will wait for the given time period before failing if unable to obtain the necessary lock.

The logs in the failing C driver tests for the stepdown test show a jump of 5 seconds, indicating it is blocking for the full 5 seconds before returning an error.

And testing with the shell, if I lock with `db.fsyncLock()`, then attempt to stepdown with `db.adminCommand({replSetStepDown: 30, force: true})`, the `replSetStepDown` will block for 30 seconds before failing with the same error. If I `db.fsyncUnlock()` while the `replSetStepDown` is blocked, it succeeds.

I'm not sure if the manual's description of the time period of `replSetStepDown` lines up with my understanding. https://docs.mongodb.com/manual/reference/command/replSetStepDown/

> The number of seconds to step down the primary, during which time the stepdown member is ineligible for becoming primary. If you specify a non-numeric value, the command uses 60 seconds.

I interpret that as: "the maximum amount of time to attempt to step down the primary, during which it is ineligible for re-election". So I _think_ we can rely on that as a timeout.

Reply:
That would also be a viable solution but increasing the stepdown time makes subsequent tests (back-to-back test runs) take longer. I prefer the retry solution because it does not slow down the test suite unnecessarily.

Reply:
Oh good point, I'll make the proposed retry change in libmongoc to validate.

Reply:
Testing with a 10 second retry at least works in libmongoc. But this test does not fail very often to begin with, so I'm not sure if that's much assurance.